### ISSUE TITLE
`sample` method - proper handling of models with no parameters

### ISSUE THREAD

**Issue by docsteveharris** [stan-dev/cmdstanpy#425]
#### Summary:

I get the following error when trying to use the generated quantities block to do prior predictive checks.

```py
/opt/homebrew/Caskroom/miniforge/base/envs/stan-base/lib/python3.8/site-packages/cmdstanpy/utils.py in scan_metric(fd, config_dict, lineno)
    714     lineno += 1
    715     if not line == '# Adaptation terminated':
--> 716         raise ValueError(
    717             'line {}: expecting metric, found:\n\t "{}"'.format(lineno, line)
    718         )

ValueError: line 1045: expecting metric, found:
	 "#"
```

I think I am following this [advice](https://mc-stan.org/docs/2_27/stan-users-guide/prior-predictive-checks.html#coding-prior-predictive-checks-in-stan) in the Stan users' guide.

```stan
data {
  int<lower = 0> N;
  vector[N] x;
}
generated quantities {
  real alpha = normal_rng(0, 1);
  real beta = normal_rng(0, 1);
  real y_sim[N] = poisson_log_rng(alpha + beta * x);
}
```
I have asked for help on the Stan user's forum without success so far. I wonder though if this is to do with the **CmdStanPy** implementation since the error is arising from `utils.scan_metric`?

#### Description:

Full details and a MWE are provided on the stan forums.
See https://discourse.mc-stan.org/t/implementing-prior-predictive-checks-in-stan/23815

#### Additional Information:
I can get a similar model to compile using the transformed parameters block instead.
I'm posting here since there are fewer Python than R users there are on the Stan forum, and if this is _not_ a problem with **CmdStanPy** then I can go back to those forums confirming that I've checked here.

#### Current Version:

```
Python implementation: CPython
Python version       : 3.8.10
IPython version      : 7.25.0

matplotlib: 3.4.2
seaborn   : 0.11.1
cmdstanpy : 0.9.76
pandas    : 1.3.1
arviz     : 0.11.2
numpy     : 1.21.1
```

Many thanks

**Comment by mitzimorris** [comment]
you don't need to run standalone generated quantities here - you can just run the sampler, because your generated quantities block doesn't reference any model parameters.  (that said, this is an interesting example of a likely user error, and we should add it to CmdStan and CmdStanPy test set and figure out good error messages.)

the error you're getting indicates that there's a problem with file or files which are the `mcmc_sample` argument to the  `generate` method.   what is that file?  given that you don't need any kind of input sample here, what did you pass in?

we've just done a bunch of PR's that will improve the workflow for running standalone generated quantities, the  the resulting CmdStanGQ object - https://github.com/stan-dev/cmdstanpy/pull/413, and as soon as the documentation has been properly updated, we will do a new CmdStanPy release.

**Comment by WardBrian** [comment]
I think this problem actually is with `sample` -- I was able to recreate the problem using his model and 
`ppc.sample({'N':20, 'x':np.zeros(20)})`

**Comment by mitzimorris** [comment]
I think this is because CmdStan now runs `fixed_params=TRUE` whenever a model has no parameters.
the `CmdStanMCMC` class isn't aware of this, and fails on initialization.  damn.  we definitely need a unit test for this use case.

**Comment by WardBrian** [comment]
I think that the csv format is simply different if there is no model - for example


```stan
data {
  int<lower = 0> N;
  vector[N] x;
}
parameters {
    real dummy;
}
model {
    dummy ~ uniform(-1,1);
}
generated quantities {
  real alpha = normal_rng(0, 1);
  real beta = normal_rng(0, 1);
  real y_sim[N] = poisson_log_rng(alpha + beta * x);
}
```

works (and I suppose is a workaround for @docsteveharris while we investigate this)

 The only appreciable difference in the output that I can see is the addition of 

```
# Adaptation terminated
# Step size = 0.815036
# Diagonal elements of inverse mass matrix:
# 3.55155
```

**Comment by WardBrian** [comment]
Ah, setting `fixed_param=True` does the immediate error, but it breaks things like `.draws_xr()` because I decided to include `num_unconstrained_params` in the metadata and that doesn't exist now. Whoops

**Comment by mitzimorris** [comment]
> I think that the csv format is simply different if there is no model

exactly - there's neither warmup draws nor adaptation information - this isn't documented here - should be: https://mc-stan.org/docs/2_27/cmdstan-guide/stan-csv.html#mcmc-sampler-csv-output

**Comment by mitzimorris** [comment]
the fix for this involves twiddling with the logic for CmdStanMCMC method
`_validate_csv_files`
I can take a crack at this over the weekend.

**Comment by mitzimorris** [comment]
the fix for this involves twiddling with the logic for CmdStanMCMC method `_validate_csv_files`
and the logic for attribute `_is_fixed_param`.    I can see how to fix this - once this value is properly set, we can update `draws_xr` logic accordingly.

I could do this over the weekend.

**Comment by WardBrian** [comment]
> exactly - there's neither warmup draws nor adaptation information - this isn't documented here - should be: https://mc-stan.org/docs/2_27/cmdstan-guide/stan-csv.html#mcmc-sampler-csv-output

I'm seeing _"Note that when the sampler is run with arguments algorithm=fixed_param, this section will be missing."_ at the bottom of the section on "Comment rows for adaptation", but it certainly could be clearer. 

Will the fix mean `num_unconstrained_params` is populated always, or should I look into making the `draws_xr` function more robust?

**Comment by mitzimorris** [comment]
technically, if `_is_fixed_params`, then `num_unconstrained_params = 0`.
so all we need is a little conditional logic - the should be straightforward unless there's something I don't understand about xarray - is there something I'm missing?

**Comment by WardBrian** [comment]
No, I just wasn't sure of that fact -- sounds good to me

**Comment by mitzimorris** [comment]
two options:

- hack CmdStan in order to get CSV output to record that algorithm is actually 'fixed_params' (not HMC)
- compilcate scan CSV logic by alot.

much prefer the former to the latter.

**Comment by WardBrian** [comment]
Is it safe to assume that cmdstanr exhibits a similar error, or have they take route 2?

**Comment by mitzimorris** [comment]
not sure, but now I see that CmdStan will issue a warning, so I can check for that warning in the outputs and this will solve the problem.

however, `from_csv` method will need an extra flag because it won't be able to handle parsing the CSV output file by itself.

**Comment by mitzimorris** [comment]
there's a ton of logic in CmdStanPy that can be eliminated, due to changes made here: https://github.com/stan-dev/cmdstanpy/pull/346, which changed parsing of the CSV header file - previously, only non-default config was recorded.  now all config is recorded.  therefore we always know specified number of warmup, sampling iters, ditto save_warmup, and thin.

since the expected number of warmup draws are known, checking the sampler files becomes much easier.
if the algorithm is 'fixed_param', don't check for metric.  if algorithm is 'hmc', check for metric, but if not found, change algorithm to 'fixed_param'.   the attribute `_is_fixed_param` can be eliminated entirely, in favor metadata config dict.

**Comment by WardBrian** [comment]
Do you think this is reasonable to get into the next version (the one with the GQ refactor), or is it a large enough undertaking that it should wait?

**Comment by mitzimorris** [comment]
the GQ refactor took about 2 x's as long as I thought it would.  that's par for the course, I guess.  this cleanup is perhaps similar.  the branch I have now will fix the problem at hand, and all it needs it unit tests, so I think I should write the unit tests to test the hacky logic and get that in ASAP.

**Comment by mitzimorris** [comment]
@WardBrian:  why does xarray care about num_unconstrained_parameters?  the Stan output is in terms of constrained parameters?
if you output the stepsize and metric, num_unconstrained_parameters is relevant, because its the size of the vector/matrix.

**Comment by WardBrian** [comment]
I put it in the attributes as part of the subset of metadata I decided to include with the dataset, completely arbitrarily. We can just as well not include it

**Comment by WardBrian** [comment]
Resolved by #427 for now, will be revisited in the future for a cleaner solution

**Comment by docsteveharris** [comment]
Just wanted to drop in at the end of this and say a massive thanks. 
Looking forward to trying out the fix.
